src/hostnet/forwards.ml

```diff
@@ -0,0 +1,487 @@
+  type flow = { mutable remaining : Cstruct.t; flow : FLOW.flow }
+
+  let connect flow = { remaining = Cstruct.create 0; flow }
+
+  type error = FLOW.error
+
+  let pp_error = FLOW.pp_error
+
+  type write_error = FLOW.write_error
+
+  let pp_write_error = FLOW.pp_write_error
+
+  let read_some flow len =
+    let open Lwt.Infix in
+    let rec loop acc len =
+      if Cstruct.length flow.remaining = 0 && len > 0 then
+        FLOW.read flow.flow >>= function
+        | Error e -> Lwt.return (Error e)
+        | Ok (`Data buf) ->
+            flow.remaining <- buf;
+            loop acc len
+        | Ok `Eof -> Lwt.return (Ok `Eof)
+      else if Cstruct.length flow.remaining < len then (
+        let take = flow.remaining in
+        flow.remaining <- Cstruct.create 0;
+        loop (take :: acc) (len - Cstruct.length take))
+      else
+        let take, leave = Cstruct.split flow.remaining len in
+        flow.remaining <- leave;
+        Lwt.return @@ Ok (`Data (List.rev (take :: acc)))
+    in
+    loop [] len
+
+  let read flow =
+    if Cstruct.length flow.remaining > 0 then (
+      let result = flow.remaining in
+      flow.remaining <- Cstruct.create 0;
+      Lwt.return @@ Ok (`Data result))
+    else FLOW.read flow.flow
+
+  let write flow = FLOW.write flow.flow
+  let writev flow = FLOW.writev flow.flow
+  let close flow = FLOW.close flow.flow
+  let shutdown_write flow = FLOW.shutdown_write flow.flow
+  let shutdown_read flow = FLOW.shutdown_read flow.flow
+end
+
+module Handshake (FLOW : Read_some) = struct
+  (* A Message is a buffer prefixed with a uint16 length field. *)
+  module Message = struct
+    open Lwt.Infix
+
+    let pp_error ppf = function
+      | `Flow e -> FLOW.pp_error ppf e
+      | `Eof -> Fmt.string ppf "EOF while reading handshake"
+
+    let read flow =
+      FLOW.read_some flow 2 >>= function
+      | Error e -> Lwt.return (Error (`Flow e))
+      | Ok `Eof -> Lwt.return (Error `Eof)
+      | Ok (`Data bufs) -> (
+          let buf = Cstructs.to_cstruct bufs in
+          let len = Cstruct.LE.get_uint16 buf 0 in
+          FLOW.read_some flow len >>= function
+          | Error e -> Lwt.return (Error (`Flow e))
+          | Ok `Eof -> Lwt.return (Error `Eof)
+          | Ok (`Data bufs) -> Lwt.return (Ok (Cstructs.to_cstruct bufs)))
+
+    let write flow t =
+      let len = Cstruct.create 2 in
+      Cstruct.LE.set_uint16 len 0 (Cstruct.length t);
+      FLOW.writev flow [ len; t ]
+  end
+
+  module Request = struct
+    type t = {
+      protocol : Protocol.t;
+      src_ip : Ipaddr.t;
+      src_port : int;
+      dst_ip : Ipaddr.t;
+      dst_port : int;
+    }
+
+    open Ezjsonm
+
+    let of_json j =
+      let protocol = Protocol.of_json @@ find j [ "protocol" ] in
+      let src_ip =
+        match Ipaddr.of_string @@ get_string @@ find j [ "src_ip" ] with
+        | Error (`Msg m) ->
+            raise (Parse_error (j, "src_ip should be an IPv4 address: " ^ m))
```

Comment:
Doesn't `Ipaddr.of_string` also parse V6 addresses? Dont you need `Ipaddr.V4.of_string` here?

Reply:
Ah, I think in this case the accompanying error message is wrong -- I've been trying to use `Ipaddr.t` more throughout in preparation for supporting IPv6